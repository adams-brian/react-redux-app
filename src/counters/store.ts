import { IAction } from '../store';

export interface IState {
  counters: number[];
}

/* tslint:disable:max-classes-per-file */
abstract class ActionWithNumberPayload implements IAction {
  public readonly type: string;
  constructor(public payload: number) {}
}
export const INCREMENT = '[Counters] INCREMENT';
class Increment extends ActionWithNumberPayload {
  public readonly type = INCREMENT;
}
export const DECREMENT = '[Counters] DECREMENT';
class Decrement extends ActionWithNumberPayload {
  public readonly type = DECREMENT;
}
export const RESET = '[Counters] RESET';
class Reset extends ActionWithNumberPayload {
  public readonly type = RESET;
}
export const ADD_COUNTER = '[Counters] ADD COUNTER';
class AddCounter implements IAction {
  public readonly type = ADD_COUNTER;
}
export const REMOVE_COUNTER = '[Counters] REMOVE COUNTER';
class RemoveCounter extends ActionWithNumberPayload {
  public readonly type = REMOVE_COUNTER;
}
export const COUNTERS_UPDATED = '[Counters] COUNTERS UPDATED';
class CountersUpdated implements IAction {
  public readonly type = COUNTERS_UPDATED;
  constructor(public payload: number[]) {}
}
export type CountersAction = Increment | Decrement | Reset | AddCounter | RemoveCounter | CountersUpdated;

export const actionCreators = {
  addCounter: () => Object.assign({}, new AddCounter()),
  countersUpdated: (counters: number[]) => Object.assign({}, new CountersUpdated(counters)),
  decrement: (index: number) => Object.assign({}, new Decrement(index)),
  increment: (index: number) => Object.assign({}, new Increment(index)),
  removeCounter: (index: number) => Object.assign({}, new RemoveCounter(index)),
  reset: (index: number) => Object.assign({}, new Reset(index)),
};

function countersReducer(state: number[] = [], action: CountersAction) {
  switch (action.type) {
    case INCREMENT:
      return [
        ...state.slice(0, action.payload),
        state[action.payload] + 1,
        ...state.slice(action.payload + 1)
      ];
    case DECREMENT:
      return [
        ...state.slice(0, action.payload),
        state[action.payload] - 1,
        ...state.slice(action.payload + 1)
      ];
    case RESET:
      return [
        ...state.slice(0, action.payload),
        0,
        ...state.slice(action.payload + 1)
      ];
    case ADD_COUNTER:
      return [...state, 0];
    case REMOVE_COUNTER:
      return [
        ...state.slice(0, action.payload),
        ...state.slice(action.payload + 1)
      ];
    case COUNTERS_UPDATED:
      return [...action.payload];
    default:
      return state;
  }
}

export const reducers = {
  counters: countersReducer
}
