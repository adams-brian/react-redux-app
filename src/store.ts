import { LocationChangeAction } from 'react-router-redux';

export interface IUser { 
  _id: string;
  firstname: string;
  lastname: string;
}

export interface IAppState {
  loading: boolean;
  counters: number[];
  users: IUser[];
}

interface IAction {
  type: string;
}

/* tslint:disable:max-classes-per-file */
export const START_LOADING = '[App] START LOADING';
class StartLoading implements IAction {
  public readonly type = START_LOADING;
}
export const DONE_LOADING = '[App] DONE LOADING';
class DoneLoading implements IAction {
  public readonly type = DONE_LOADING;
}
type LoadingAction = StartLoading | DoneLoading;

abstract class ActionWithNumberPayload implements IAction {
  public readonly type: string;
  constructor(public payload: number) {}
}
export const INCREMENT = '[Counters] INCREMENT';
class Increment extends ActionWithNumberPayload {
  public readonly type = INCREMENT;
}
export const DECREMENT = '[Counters] DECREMENT';
class Decrement extends ActionWithNumberPayload {
  public readonly type = DECREMENT;
}
export const RESET = '[Counters] RESET';
class Reset extends ActionWithNumberPayload {
  public readonly type = RESET;
}
export const ADD_COUNTER = '[Counters] ADD COUNTER';
class AddCounter implements IAction {
  public readonly type = ADD_COUNTER;
}
export const REMOVE_COUNTER = '[Counters] REMOVE COUNTER';
class RemoveCounter extends ActionWithNumberPayload {
  public readonly type = REMOVE_COUNTER;
}
export const COUNTERS_UPDATED = '[Counters] COUNTERS UPDATED';
class CountersUpdated implements IAction {
  public readonly type = COUNTERS_UPDATED;
  constructor(public payload: number[]) {}
}
type CounterAction = Increment | Decrement | Reset | AddCounter | RemoveCounter | CountersUpdated;

export const CREATE_USER = '[Users] CREATE';
export class CreateUser implements IAction {
  public readonly type = CREATE_USER;
  constructor(public payload: IUser) {}
}
export const UPDATE_USER = '[Users] UPDATE';
class UpdateUser implements IAction {
  public readonly type = UPDATE_USER;
  constructor(public payload: IUser) {}
}
export const DELETE_USER = '[Users] DELETE';
export class DeleteUser implements IAction {
  public readonly type = DELETE_USER;
  constructor(public payload: string) {}
}
export const USERS_UPDATED = '[Users] USERS UPDATED';
class UsersUpdated implements IAction {
  public readonly type = USERS_UPDATED;
  constructor(public payload: IUser[]) {}
}
type UserAction = CreateUser | UpdateUser | DeleteUser | UsersUpdated;

export const actionCreators = {
  doneLoading: () => Object.assign({}, new DoneLoading()),
  startLoading: () => Object.assign({}, new StartLoading()),

  addCounter: () => Object.assign({}, new AddCounter()),
  countersUpdated: (counters: number[]) => Object.assign({}, new CountersUpdated(counters)),
  decrement: (index: number) => Object.assign({}, new Decrement(index)),
  increment: (index: number) => Object.assign({}, new Increment(index)),
  removeCounter: (index: number) => Object.assign({}, new RemoveCounter(index)),
  reset: (index: number) => Object.assign({}, new Reset(index)),

  createUser: (user: IUser) => Object.assign({}, new CreateUser(user)),
  deleteUser: (id: string) => Object.assign({}, new DeleteUser(id)),
  updateUser: (user: IUser) => Object.assign({}, new UpdateUser(user)),
  usersUpdated: (users: IUser[]) => Object.assign({}, new UsersUpdated(users)),
};

function loadingReducer(state: boolean = false, action: LoadingAction) {
  switch (action.type) {
    case START_LOADING:
      return true;
    case DONE_LOADING:
      return false;
    default:
      return state;
  }
}

function countersReducer(state: number[] = [], action: CounterAction) {
  switch (action.type) {
    case INCREMENT:
      return [
        ...state.slice(0, action.payload),
        state[action.payload] + 1,
        ...state.slice(action.payload + 1)
      ];
    case DECREMENT:
      return [
        ...state.slice(0, action.payload),
        state[action.payload] - 1,
        ...state.slice(action.payload + 1)
      ];
    case RESET:
      return [
        ...state.slice(0, action.payload),
        0,
        ...state.slice(action.payload + 1)
      ];
    case ADD_COUNTER:
      return [...state, 0];
    case REMOVE_COUNTER:
      return [
        ...state.slice(0, action.payload),
        ...state.slice(action.payload + 1)
      ];
    case COUNTERS_UPDATED:
      return [...action.payload];
    default:
      return state;
  }
}

const usersReducer = (state: IUser[] = [], action: UserAction) => {
  switch (action.type) {
    case CREATE_USER:
      return [...state, action.payload];
    case UPDATE_USER:
      return state.map(user => user._id === action.payload._id ? Object.assign({}, action.payload) : user);
    case DELETE_USER:
      return state.filter(user => user._id !== action.payload);
    case USERS_UPDATED:
      return [...action.payload];
    default:
      return state;
  }
};

export const reducers = {
  counters: countersReducer,
  loading: loadingReducer,
  users: usersReducer
};

export type AppAction = LoadingAction | CounterAction | UserAction | LocationChangeAction;
