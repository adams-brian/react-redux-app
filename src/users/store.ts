import { Action, combineReducers } from 'redux';

export interface IUser {
  _id: string;
  firstname: string;
  lastname: string;
}

export interface IState {
  users: {
    error: string;
    list: IUser[];
    loaded: boolean;
  }
}

/* tslint:disable:max-classes-per-file */
export const CREATE_USER = '[Users] CREATE';
export class CreateUser implements Action {
  public readonly type = CREATE_USER;
  constructor(public payload: IUser) { }
}
export const UPDATE_USER = '[Users] UPDATE';
export class UpdateUser implements Action {
  public readonly type = UPDATE_USER;
  constructor(public payload: IUser) { }
}
export const DELETE_USER = '[Users] DELETE';
export class DeleteUser implements Action {
  public readonly type = DELETE_USER;
  constructor(public payload: string) { }
}
export const USERS_UPDATED = '[Users] USERS UPDATED';
class UsersUpdated implements Action {
  public readonly type = USERS_UPDATED;
  constructor(public payload: IUser[]) { }
}
export const LOAD_USERS = '[Users] LOAD USERS';
export class LoadUsers implements Action {
  public readonly type = LOAD_USERS;
}
export const USERS_ERROR = '[Users] USERS ERROR';
export class UsersError implements Action {
  public readonly type = USERS_ERROR;
  constructor(public payload: Error) { }
}
export const USERS_LOADED = '[Users] USERS LOADED';
export class UsersLoaded implements Action {
  public readonly type = USERS_LOADED;
}
export type UsersAction = CreateUser | UpdateUser | DeleteUser | UsersUpdated | LoadUsers | UsersError | UsersLoaded;

export const actionCreators = {
  createUser: (user: IUser) => Object.assign({}, new CreateUser(user)),
  deleteUser: (id: string) => Object.assign({}, new DeleteUser(id)),
  loadUsers: () => Object.assign({}, new LoadUsers()),
  updateUser: (user: IUser) => Object.assign({}, new UpdateUser(user)),
  usersError: (err: Error) => Object.assign({}, new UsersError(err)),
  usersLoaded: () => Object.assign({}, new UsersLoaded()),
  usersUpdated: (users: IUser[]) => Object.assign({}, new UsersUpdated(users)),
};

export const listReducer = (state: IUser[] = [], action: UsersAction) => {
  switch (action.type) {
    case CREATE_USER:
      return [...state, action.payload];
    case UPDATE_USER:
      return state.map(user => user._id === action.payload._id ? Object.assign({}, action.payload) : user);
    case DELETE_USER:
      return state.filter(user => user._id !== action.payload);
    case USERS_UPDATED:
      return [...action.payload];
    default:
      return state;
  }
};

export const loadedReducer = (state: boolean = false, action: UsersAction) => {
  switch (action.type) {
    case USERS_LOADED:
      return true;
    default:
      return state;
  }
}

export const errorReducer = (state: string = '', action: UsersAction) => {
  switch (action.type) {
    case USERS_ERROR:
      return action.payload.message;
    default:
      return state;
  }
}

export const reducers = {
  users: combineReducers({
    error: errorReducer,
    list: listReducer,
    loaded: loadedReducer
  })
};
