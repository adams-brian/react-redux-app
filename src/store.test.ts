import { actionCreators, DONE_LOADING, LoadingAction, reducers, START_LOADING } from "./store";

describe('store', () => {
  
  describe('actions', () => {

    it('should export action types', () => {
      expect(START_LOADING).toMatchSnapshot();
      expect(DONE_LOADING).toMatchSnapshot();
    });

  });

  describe('actionCreators', () => {
    
    it('should export actionCreators', () => {
      expect(actionCreators.doneLoading()).toMatchSnapshot();
      expect(actionCreators.startLoading()).toMatchSnapshot();
    });

  });

  describe('reducer', () => {

    it('should return true when starting to load', () => {
      expect(reducers.loading(undefined, actionCreators.startLoading())).toBe(true);
      expect(reducers.loading(false, actionCreators.startLoading())).toBe(true);
      expect(reducers.loading(true, actionCreators.startLoading())).toBe(true);
    });

    it('should return false when done loading', () => {
      expect(reducers.loading(undefined, actionCreators.doneLoading())).toBe(false);
      expect(reducers.loading(false, actionCreators.doneLoading())).toBe(false);
      expect(reducers.loading(true, actionCreators.doneLoading())).toBe(false);
    });

  });

});