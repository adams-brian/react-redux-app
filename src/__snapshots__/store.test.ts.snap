// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store actionCreators should export actionCreators 1`] = `
Object {
  "type": "[App] DONE LOADING",
}
`;

exports[`store actionCreators should export actionCreators 2`] = `
Object {
  "type": "[App] START LOADING",
}
`;

exports[`store actions should export action types 1`] = `"[App] START LOADING"`;

exports[`store actions should export action types 2`] = `"[App] DONE LOADING"`;
